keps/sig-node/3619-supplemental-groups-policy/README.md

```diff
@@ -0,0 +1,912 @@
+  containers:
+    # Expected output: "uid=1000(alice) gid=1000(alice) groups=1000(alice),60000"
+    # NOTE: "group-in-image" is not included here 
+    #       because groups defined in /etc/group should not be attached
+    #       when gids is specified in runAsGroup
+  - image: the-image-above
+    sh: ["id"] 
+```
+
+However, the current combination with Kubernetes and major containter runtimes(at least containerd and cri-o) outputs(See [here](https://github.com/pfnet-research/strict-supplementalgroups-container-runtime/tree/reproduce-bypass-supplementalgroups) for more detailed reproduction code) includes "group-in-image" group of the first container process.
+
+```console
+uid=1000(alice) gid=1000(alice) groups=1000(alice),50000(group-in-image),60000
+```
+
+## Motivation
+
+As described above, gids attached to the first container process is not OCI image spec compliant when `runAsGroup` is specified in `SecurityContext`.
+
+Moreover, this causes security considerations as follows. When a cluster enforces some security policy for pods that protects the value of `RunAsGroup` and `SupplementalGroups`, the effect of its enforcement is limited, i.e., cluster users can easily bypass the policy enforcement just by using a custom image. If such a bypass happened, it would be unexpected behavior for most cluster administrators because the enforcement is almost useless. Moreover, the bypass will cause unexpected file access permission. In some use cases, the unexpected file access permission will be a security concern. For example, using `hostPath` volumes could be a severe problem because UID/GIDs matter in accessing files/directories in the volumes.
+
+Kubernetes provides no API surface to prevent this bypass, although it could sometimes lead to a security concern. Because the behavior is implemented in CRI implementations actually, To mitigate this, the cluster administrators will need to deploy a custom low-level container runtime(e.g., [pfnet-research/strict-supplementalgroups-container-runtime](https://github.com/pfnet-research/strict-supplementalgroups-container-runtime)) that modifies OCI container runtime spec(`config.json`) produced by CRI implementations (e.g., containerd, cri-o). A custom `RuntimeClass` would be introduced for it. Nevertheless, It would be an extra operational burden for cluster administrators.
+
+Thus, this KEP proposes to offer a new API field named `SupplementalGroupsPolicy` that enables OCI-compliant behavior for supplemental groups to the first container process. The new API allows cluster administrators to deploy security policies that protect the `SupplementalGroupsPolicy` field in the cluster to avoid the unexpected bypass of `SupplementalGroups` described above.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- To Provide a new API field to control exactly which groups the container process belongs to
```

Comment:
how strong do we believe this must be per-Pod or just a global setting on a Runtime? As long as there is a way to check that current workload is not affected, switching a global flag sounds like a better solution long term.

Or there is a use for this setting after every implementation has migrated to the new behavior?

Reply:
How do you check it is not affected?

Reply:
we may need to get this from runtime. If changing it for the Pod and seeing if something breaks the only option for end users - it sounds bad. Also I'm not sure if this field will be useful long term. Do you think we will need it?

Reply:
Ultimately I am trying to understand if there are set of Goals we can set for 1.27 to get this KEP approved for 1.27 and decide on specifics later during implementation.

Reply:
Well, we either need a breaking change or an API, I think. And APIs are forever.

I am ready to ACK an API if we can get the semantic roughly described including how to know if this might impact you (status).

Reply:
I added `ContainerStatus.User` proposal in the KEP for users/admins to know the actual identities attached to the container processes in 8b3e39b07b8a436aa4070e0a9d96f36ed6a16760. PTAL.